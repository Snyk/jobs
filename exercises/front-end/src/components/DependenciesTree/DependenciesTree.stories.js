import '~/assets/scss/styles.scss';
import DependenciesTree from './DependenciesTree';

import * as depGraph from './depGraph.fixture';

export default {
  title: 'Library/DependenciesTree',
  component: DependenciesTree,
};

export const Default = () => ({
  components: { DependenciesTree },
  provide: {
    project: { type: 'npm' },
  },
  props: {
    depGraph: {
      default: depGraph,
    },
    forceCollapsedState: {
      default: 'expanded',
    },
  },
  template: `<DependenciesTree v-bind="$props" />`,
});
