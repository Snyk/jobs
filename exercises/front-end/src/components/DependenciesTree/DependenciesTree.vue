<!-- cSpell:words renderable  -->
<template>
  <ul :class="computedClasses" class="vue--tree">
    <li :data-snyk-test="testString" class="vue--tree__item">
      <div
        :tabindex="tabindex"
        class="vue--tree__package"
        @click="handleCollapse"
        @keyup.enter="handleCollapse"
      >
        <a
          :href="testUrl"
          :title="packageTitle"
          class="vue--tree__link"
          @click.stop
        >
          {{ packageTitle }}
        </a>
        <BaseLoadingSpinner
          v-if="isRoot && collapsing"
          class="vue--tree__spinner"
          size="small"
        />
      </div>
      <template v-if="isRoot || !collapsed">
        <ProjectDependenciesTree
          v-for="(nodeDependency, i) in Object.values(pkg.dependencies)"
          :key="i"
          :dep-graph="nodeDependency"
          :parent-nest-length="Object.keys(pkg.dependencies).length"
          :index="i"
          :force-collapsed-state="collapseState"
          :is-root="false"
        />
      </template>
    </li>
  </ul>
</template>

<script>
import BaseLoadingSpinner from '~/components/BaseLoadingSpinner/BaseLoadingSpinner';

import { isInList } from '~/lib/prop-validators';

/**
 * The `Tree` component renders a Snyk dependency tree with associated vulnerabilities and license issues.
 */
export default {
  name: 'ProjectDependenciesTree',

  status: 'ready',

  components: {
    BaseLoadingSpinner,
  },

  props: {
    /**
     * RenderableGraph, i.e. hydrated with vuln data and indices.
     * Use `toRenderableGraph` in `/frontend/src/lib/dep-graph-utils.js`
     */
    depGraph: {
      type: Object,
      required: true,
    },
    /**
     * Point in dep graph being rendered.
     */
    isRoot: {
      type: Boolean,
      default: true,
    },
    /**
     * @ignore
     */
    index: {
      type: Number,
      default: 0,
    },
    /**
     * @ignore
     */
    parentNestLength: {
      type: Number,
      required: false,
      default: 0,
    },
    /**
     * Hide license issue labels.
     */
    suppressLicenseLabels: {
      type: Boolean,
      default: false,
    },
    /**
     * Hide issue labels.
     */
    suppressIssueLabels: {
      type: Boolean,
      default: false,
    },
    /**
     * Force collapsed state for child dependencies.
     */
    forceCollapsedState: {
      type: String,
      default: 'expanded',
      validator: isInList(['expanded', 'collapsed']),
    },
  },

  data() {
    return {
      collapseState: this.forceCollapsedState,
      collapsed: false,
      collapsing: false,
    };
  },

  computed: {
    pkg() {
      return this.depGraph;
    },
    dependencies() {
      return this.pkg.dependencies;
    },
    name() {
      return this.pkg.name;
    },
    version() {
      return this.pkg.version;
    },
    computedClasses() {
      return {
        'vue--tree--collapsible': this.isCollapsible,
        'vue--tree--collapsed': this.isCollapsible && this.collapsed,
        'vue--tree--last': this.last,
        'vue--tree--collapsed-all':
          this.isRoot && this.collapseState === 'collapsed',
        'vue--tree--collapsing': this.collapsing,
      };
    },
    testUrl() {
      return `https://snyk.io/test/npm/${this.name}@${this.version}`;
    },
    packageTitle() {
      const excludeVersions = ['0.0.0'];

      if (!this.name || this.name === 'project') return 'Project';

      if (!this.version) return this.name;

      const label = !excludeVersions.includes(this.version)
        ? `${this.name}@${this.version}`
        : this.name;

      return label;
    },
    nestLength() {
      if (!this.dependencies) return 0;

      return Object.keys(this.dependencies).length;
    },
    tabindex() {
      return this.isCollapsible || this.isRoot ? 0 : -1;
    },
    last() {
      return (
        (this.nestLength >= this.parentNestLength &&
          this.index + 1 == this.parentNestLength) ||
        (this.parentNestLength == this.index + 1 && this.nestLength > 0)
      );
    },
    isCollapsible() {
      return this.nestLength > 0 && this.parentNestLength !== 0;
    },
    testString() {
      const childDepCount = Object.keys(this.dependencies).length
        ? Object.keys(this.dependencies).length
        : 0;
      return `${this.name}#${childDepCount}`;
    },
  },

  watch: {
    forceCollapsedState: {
      handler() {
        this.collapseState = this.forceCollapsedState;
        this.handleForceCollapsedState();
      },
      deep: true,
    },
  },

  created() {
    this.handleForceCollapsedState();
  },

  methods: {
    handleForceCollapsedState() {
      if (this.collapseState === 'expanded') {
        this.collapsed = false;
      }
      if (this.collapseState === 'collapsed') {
        this.collapsed = true;
      }
    },
    handleCollapse() {
      if (!this.isRoot) {
        this.toggleCollapsed();
      }
    },
    toggleCollapsed() {
      this.collapsed = !this.collapsed;
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--tree {
  $self: &;

  border-left: none !important; // Override legacy
  display: block;
  list-style: none;
  margin: 0;
  padding: 0 0 0 space(m);
  position: relative;
  width: 100%;

  &::before {
    background-color: color(neutral, 72);
    content: '';
    display: block;
    height: calc(100% + 12px);
    left: space(s);
    position: absolute;
    top: -31px;
    width: 2px;
  }

  &:first-child {
    padding-left: 0;

    &::before {
      display: none;
    }
  }

  &:first-child > #{$self}__item:first-child > #{$self}__package {
    background-color: color(neutral, 96);
    margin-top: 0;

    &::before {
      display: none;
    }
  }

  &--collapsed-all {
    &:first-of-type > #{$self}__item:first-child > #{$self}__package {
      cursor: default;
      &::after {
        display: none;
      }
    }
  }

  &--collapsible {
    > #{$self}__item:first-child > #{$self}__package {
      cursor: pointer;
      transition: box-shadow 0.2s ease;

      &:hover {
        box-shadow: inset 0 0 0 1px color(action), 0 0 0 1px color(action);
      }

      &::after {
        border: 1px solid color(neutral, 72);
        content: '-';
        display: block;
        height: 20px;
        font-size: 1.2em;
        line-height: 1.3;
        position: absolute;
        text-align: center;
        right: 6px;
        top: 8px;
        width: 20px;

        @include media-query(m) {
          line-height: 1.1;
        }
      }
    }

    &#{$self}--collapsed {
      > #{$self}__item:first-child > #{$self}__package {
        &::after {
          content: '+';
          line-height: 1.3;

          @include media-query(m) {
            line-height: 1.2;
          }
        }
      }
    }

    &#{$self}--collapsing {
      > #{$self}__item:first-child > #{$self}__package {
        &::after {
          display: none;
        }
      }
    }
  }

  &--last {
    &::before {
      height: 51px;
    }
  }

  &__spinner {
    background-color: color(neutral, 84);
    height: 34px;
    margin-top: 13px;
    position: absolute;
    transform: translateY(-50%);
    right: 1px;
    width: 34px;
    z-index: 10;
  }

  &__item {
    padding: 0;
    margin: 0;

    &--hidden {
      display: none;
    }
  }

  &__package {
    align-items: center;
    background-color: color(ui, white);
    border-radius: 2px;
    box-shadow: inset 0 0 0 1px color(neutral, 72), 0 0 0 2px transparent;
    display: flex;
    font-size: rem(14px);
    line-height: 1.6;
    font-variant: proportional-nums;
    height: 38px;
    margin: space(s) 0;
    padding: space(s) * 0.5 space(xxl) space(s) * 0.5 space(s);
    position: relative;
    z-index: 2;

    @include media-query(m) {
      font-size: 1rem;
    }

    &::before {
      background-color: color(neutral, 72);
      content: '';
      display: block;
      height: 2px;
      left: -#{space(s)};
      margin-top: -1px;
      position: absolute;
      top: 50%;
      width: space(s);
      z-index: 1;
    }
  }

  &__expand-all-spinner {
    float: right;
  }

  &__severity {
    left: -#{space(s) * 0.5};
    margin-right: space(xxxs);
    position: relative;
    flex-shrink: 0;
  }

  &__status {
    display: flex;
    color: color(base, inverted);
    font-size: 0.875rem;
    padding: space(s);

    @include media-query(m) {
      font-size: 1rem;
      padding-top: 0;
      padding-bottom: 0;
    }

    .folding-cube {
      margin-right: space(s);
    }

    span {
      line-height: 1.9;
    }
  }

  &__link,
  &__label {
    display: inline-block;
    overflow: hidden;
    position: relative;
    text-overflow: ellipsis;
    white-space: nowrap;
    top: 2px;
  }

  &__badge {
    margin-left: 1em;
  }

  &__link {
    border: none !important; // Override legacy styles.

    &:hover {
      color: color(action);
      text-decoration: underline;
    }
  }
}
</style>
