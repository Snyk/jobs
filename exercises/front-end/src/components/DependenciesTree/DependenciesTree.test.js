import { mount, shallowMount } from '@vue/test-utils';
import DependenciesTree from './DependenciesTree';
import * as depGraph from './depGraph.fixture';

jest.useFakeTimers();

describe('DependenciesTree', () => {
  it('renders as expected', () => {
    const wrapper = mount(DependenciesTree, {
      propsData: {
        depGraph,
        nodeId: 'root-node',
        forceCollapsedState: 'collapsed',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('Defaults', () => {
    const wrapper = shallowMount(DependenciesTree, {
      propsData: { depGraph },
    });

    it('defaults to no forced collapsed state', () => {
      expect(wrapper.props('forceCollapsedState')).toEqual('expanded');
    });
  });

  describe('Behaviour', () => {
    it('handles the collapse state for children', async () => {
      const wrapper = mount(DependenciesTree, {
        propsData: {
          depGraph,
          forceCollapsedState: 'expanded',
          isRoot: false,
        },
      });

      jest.spyOn(wrapper.vm, 'toggleCollapsed');

      wrapper.vm.handleCollapse();

      expect(wrapper.vm.toggleCollapsed.mock.calls.length).toBe(1);
    });

    it('toggles collapsed state', () => {
      const wrapper = mount(DependenciesTree, {
        propsData: {
          depGraph,
          forceCollapsedState: 'expanded',
        },
      });
      wrapper.vm.toggleCollapsed();
      expect(wrapper.vm.collapsed).toBe(true);

      wrapper.vm.toggleCollapsed();
      expect(wrapper.vm.collapsed).toBe(false);
    });
  });
});
