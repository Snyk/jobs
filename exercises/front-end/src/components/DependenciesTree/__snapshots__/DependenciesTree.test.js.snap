// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependenciesTree renders as expected 1`] = `
<ul class="vue--tree vue--tree--collapsed-all" nodeid="root-node">
  <li data-snyk-test="express#30" class="vue--tree__item">
    <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/express@4.17.1" title="express@4.17.1" class="vue--tree__link">
        express@4.17.1
      </a>
      <!---->
    </div>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="accepts#2" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/accepts@1.3.7" title="accepts@1.3.7" class="vue--tree__link">
            accepts@1.3.7
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="array-flatten#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/array-flatten@1.1.1" title="array-flatten@1.1.1" class="vue--tree__link">
            array-flatten@1.1.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="body-parser#10" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/body-parser@1.19.0" title="body-parser@1.19.0" class="vue--tree__link">
            body-parser@1.19.0
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="content-disposition#1" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/content-disposition@0.5.3" title="content-disposition@0.5.3" class="vue--tree__link">
            content-disposition@0.5.3
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="content-type#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/content-type@1.0.4" title="content-type@1.0.4" class="vue--tree__link">
            content-type@1.0.4
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="cookie#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/cookie@0.4.0" title="cookie@0.4.0" class="vue--tree__link">
            cookie@0.4.0
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="cookie-signature#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/cookie-signature@1.0.6" title="cookie-signature@1.0.6" class="vue--tree__link">
            cookie-signature@1.0.6
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="debug#1" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/debug@2.6.9" title="debug@2.6.9" class="vue--tree__link">
            debug@2.6.9
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="depd#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/depd@1.1.2" title="depd@1.1.2" class="vue--tree__link">
            depd@1.1.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="encodeurl#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/encodeurl@1.0.2" title="encodeurl@1.0.2" class="vue--tree__link">
            encodeurl@1.0.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="escape-html#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/escape-html@1.0.3" title="escape-html@1.0.3" class="vue--tree__link">
            escape-html@1.0.3
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="etag#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/etag@1.8.1" title="etag@1.8.1" class="vue--tree__link">
            etag@1.8.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="finalhandler#7" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/finalhandler@1.1.2" title="finalhandler@1.1.2" class="vue--tree__link">
            finalhandler@1.1.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="fresh#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/fresh@0.5.2" title="fresh@0.5.2" class="vue--tree__link">
            fresh@0.5.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="merge-descriptors#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/merge-descriptors@1.0.1" title="merge-descriptors@1.0.1" class="vue--tree__link">
            merge-descriptors@1.0.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="methods#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/methods@1.1.2" title="methods@1.1.2" class="vue--tree__link">
            methods@1.1.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="on-finished#1" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/on-finished@2.3.0" title="on-finished@2.3.0" class="vue--tree__link">
            on-finished@2.3.0
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="parseurl#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/parseurl@1.3.3" title="parseurl@1.3.3" class="vue--tree__link">
            parseurl@1.3.3
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="path-to-regexp#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/path-to-regexp@0.1.7" title="path-to-regexp@0.1.7" class="vue--tree__link">
            path-to-regexp@0.1.7
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="proxy-addr#2" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/proxy-addr@2.0.7" title="proxy-addr@2.0.7" class="vue--tree__link">
            proxy-addr@2.0.7
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="qs#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/qs@6.7.0" title="qs@6.7.0" class="vue--tree__link">
            qs@6.7.0
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="range-parser#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/range-parser@1.2.1" title="range-parser@1.2.1" class="vue--tree__link">
            range-parser@1.2.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="safe-buffer#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/safe-buffer@5.1.2" title="safe-buffer@5.1.2" class="vue--tree__link">
            safe-buffer@5.1.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="send#13" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/send@0.17.1" title="send@0.17.1" class="vue--tree__link">
            send@0.17.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="serve-static#4" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/serve-static@1.14.1" title="serve-static@1.14.1" class="vue--tree__link">
            serve-static@1.14.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="setprototypeof#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/setprototypeof@1.1.1" title="setprototypeof@1.1.1" class="vue--tree__link">
            setprototypeof@1.1.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="statuses#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/statuses@1.5.0" title="statuses@1.5.0" class="vue--tree__link">
            statuses@1.5.0
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree vue--tree--collapsible vue--tree--collapsed">
      <li data-snyk-test="type-is#2" class="vue--tree__item">
        <div tabindex="0" class="vue--tree__package"><a href="https://snyk.io/test/npm/type-is@1.6.18" title="type-is@1.6.18" class="vue--tree__link">
            type-is@1.6.18
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="utils-merge#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/utils-merge@1.0.1" title="utils-merge@1.0.1" class="vue--tree__link">
            utils-merge@1.0.1
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
    <ul class="vue--tree">
      <li data-snyk-test="vary#0" class="vue--tree__item">
        <div tabindex="-1" class="vue--tree__package"><a href="https://snyk.io/test/npm/vary@1.1.2" title="vary@1.1.2" class="vue--tree__link">
            vary@1.1.2
          </a>
          <!---->
        </div>
        <!---->
      </li>
    </ul>
  </li>
</ul>
`;
